Support dynamic library

melonedo opened the pull request:
By omitting `STATIC` in `add_library`, user can build either static or dynamic library on their discretion base on `BUILD_SHARED_LIBS` variable. `lua` and `luac` are also installed if built.

melonedo committed:
Support dynamic library

walterschell commented:
@melonedo Thank you for the PR. My preferred way of dealing with multiple configurations is to have multiple targets and just build them all. My way forward here would be:
- [ ] Add a second identical target for `lua_shared`. 
- [ ] Control if the targets get added with two options `LUA_ENABLE_SHARED` and `LUA_ENABLE_STATIC` both defaulted to on
- [ ] Add logic to check make sure at least one of the options is enabled
- [ ] (Optional) If we are not top level, and both options have been set, use `BUILD_SHARED_LIBS` to apply a `EXCLUDE_FROM_ALL` target property. 
- [ ] (Optional) Add an alias target `Lua::lib` and where it points is determined by the `BUILD_SHARED_LIBS` variable

No issue with installing `lua` or `luac` when built. Can do a separate PR for them or wait to do them with this one.

melonedo committed:
Adapt to suggestions

- build shared and/or static version controlled by `LUA_ENABLE_SHARED/STATIC`
- add alias Lua::Lib which is one of above, controlled by `BUILD_SHARED_LIBS`
- exclude the unneeded version if not toplevel

melonedo commented:
All targets are met.

walterschell commented:
Thank you for resubmitting! My apologies I didn't get to this last weekend, I should be able to get this done by the end of this coming weekend.

walterschell commented:
It looks like using the object library instead of listing sources breaks compilation for shared objects. Can you try using the sources instead?

melonedo committed:
Compile dynamic version from source

walterschell commented:
Closing this pull request, so I can re-open to force the checks which now include the Lua self tests to run.

walterschell commented:
Thanks for this. Sorry it took so long.
